### DISCUSSION TITLE
NextJS to have dynamic content from the server and not modified on the client side

### DISCUSSION THREAD

**Discussion by cshekhar017** [vercel/next.js#63312] [Help]
### Summary

We are exploring ways to incorporate dynamic server-rendered content in Next.js while preserving its static nature on the client side. Currently, we are using gtranslate for text translation, and we've observed that disabling client JavaScript allows the translations to function properly. However, this also leads to a loss of interactivity in our application.

We are seeking guidance on alternative approaches that would allow us to maintain static content while keeping interactive elements active, even when client JavaScript is disabled. We have experimented with unstable_runtimeJS: false in our Next.js configuration, which achieves static content but compromises interactivity.

Any suggestions or best practices on how to achieve this balance would be greatly appreciated. Thank you in advance for your assistance.

### Additional information

```js
To demonstrate the issue, we have two sample URLs:

Without Client-Side JavaScript(https://next-sample.netlify.app/withoutscript) - Shows static content without client-side interactivity, when you click on value JavaScript alert will not come

With Client-Side JavaScript(https://next-sample.netlify.app/withscript) - Demonstrates client-side interactivity, but the content becomes dynamic, when you click on value JavaScript alert will come

We have also deployed our codebase to a public repository on GitHub: https://github.com/cshekhar017/next-netlify-starter
```

### Example

https://next-sample.netlify.app/withoutscript https://next-sample.netlify.app/withscript

**Comment by icyJoseph** [comment]
It sounds like you need to use the App Router, which uses Server Components, that are rendered only server side. 

That way you can still sprinkle some interactivity right where you need it, rather than opting in the entire tree. You'd only ship code for the interactive parts as well.

**Reply by cshekhar017** [reply]
Thanks for the suggestion icyJoseph, we already tried with App Router, but it did not work with https://gtranslate.io/ text translation not worked properly

**Reply by icyJoseph** [reply]
What did not work properly. App Router is the way to go here.

**Reply by cshekhar017** [reply]
We created page using App Router, and deployed this on our site, to see if translation works with https://gtranslate.io/, page content did not translated and remain in English.

We have deployed sample code on git repo https://github.com/cshekhar017/next-netlify-starter/blob/main/app/page.js, also added a sample code for App Router, suggest if we are doing anything wrong 

```
async function getProjects() {
  const res = await fetch(`https://dummyjson.com/products`, {
    cache: "no-store",
  });
  const projects = await res.json();

  return projects;
}

export default async function Home() {
  const projects = await getProjects();
  return (
    <ul>
      {projects.products.map((project) => (
        <li key={project.id}>
          {project.title}
        </li>
      ))}
    </ul>
  );
}
```

**Reply by icyJoseph** [reply]
It looks ok, but the thing here is that I have no motivation to look up gtranslate, and figure out how it works. So if you help you'll have to show how is it that the gtranslate service does translations on the page.

**Reply by cshekhar017** [reply]
When we access url of our sample site https://hi.eximshala.com/, ideally content for this should we show in Hindi language, but these are currently showing in English,  if we do View page source then we can see content is showing in hindi as required but on frontend it is not working.

![image](https://github.com/vercel/next.js/assets/113338057/71131235-cff9-4d73-9482-19b61349a2ce)

**Reply by icyJoseph** [reply]
The problem here is hydration. The scripts underneath are changing the content, doing hydration. 

The translate framework makes some wild assumptions I suppose.
